nav2_behavior_tree/test/plugins/action/test_bt_cancel_action_node.cpp

```diff
@@ -0,0 +1,199 @@
+  : nav2_behavior_tree::BtCancelActionNode<test_msgs::action::Fibonacci>(xml_tag_name, "fibonacci",
+      conf)
+  {}
+};
+
+class BTCancelActionNodeTestFixture : public ::testing::Test
+{
+public:
+  static void SetUpTestCase()
+  {
+    node_ = std::make_shared<rclcpp::Node>("bt_cancel_action_node_test_fixture");
+    factory_ = std::make_shared<BT::BehaviorTreeFactory>();
+
+    config_ = new BT::NodeConfiguration();
+
+    // Create the blackboard that will be shared by all of the nodes in the tree
+    config_->blackboard = BT::Blackboard::create();
+    // Put items on the blackboard
+    config_->blackboard->set<rclcpp::Node::SharedPtr>(
+      "node",
+      node_);
+    config_->blackboard->set<std::chrono::milliseconds>(
+      "server_timeout",
+      std::chrono::milliseconds(20));
+    config_->blackboard->set<std::chrono::milliseconds>(
+      "bt_loop_duration",
+      std::chrono::milliseconds(10));
+
+    client_ = rclcpp_action::create_client<test_msgs::action::Fibonacci>(
+      node_, "fibonacci");
+
+    BT::NodeBuilder builder =
+      [](const std::string & name, const BT::NodeConfiguration & config)
+      {
+        return std::make_unique<FibonacciAction>(name, config);
+      };
+
+    factory_->registerBuilder<FibonacciAction>("Fibonacci", builder);
+  }
+
+  static void TearDownTestCase()
+  {
+    delete config_;
+    config_ = nullptr;
+    node_.reset();
+    action_server_.reset();
+    factory_.reset();
+  }
+
+  void TearDown() override
+  {
+    tree_.reset();
+  }
+
+  static std::shared_ptr<FibonacciActionServer> action_server_;
+  static std::shared_ptr<rclcpp_action::Client<test_msgs::action::Fibonacci>> client_;
+
+protected:
+  static rclcpp::Node::SharedPtr node_;
+  static BT::NodeConfiguration * config_;
+  static std::shared_ptr<BT::BehaviorTreeFactory> factory_;
+  static std::shared_ptr<BT::Tree> tree_;
+};
+
+rclcpp::Node::SharedPtr BTCancelActionNodeTestFixture::node_ = nullptr;
+std::shared_ptr<FibonacciActionServer>
+BTCancelActionNodeTestFixture::action_server_ = nullptr;
+std::shared_ptr<rclcpp_action::Client<test_msgs::action::Fibonacci>>
+BTCancelActionNodeTestFixture::client_ = nullptr;
+
+BT::NodeConfiguration * BTCancelActionNodeTestFixture::config_ = nullptr;
+std::shared_ptr<BT::BehaviorTreeFactory> BTCancelActionNodeTestFixture::factory_ = nullptr;
+std::shared_ptr<BT::Tree> BTCancelActionNodeTestFixture::tree_ = nullptr;
+
+TEST_F(BTCancelActionNodeTestFixture, test_ports)
+{
+  // Creating a Fibonacci and treating it as a dummy node
+  std::string xml_txt =
+    R"(
+      <root main_tree_to_execute = "MainTree" >
+        <BehaviorTree ID="MainTree">
+            <Fibonacci/>
+        </BehaviorTree>
+      </root>)";
+
+  tree_ = std::make_shared<BT::Tree>(factory_->createTreeFromText(xml_txt, config_->blackboard));
+  auto send_goal_options = rclcpp_action::Client<test_msgs::action::Fibonacci>::SendGoalOptions();
+  auto goal_msg = test_msgs::action::Fibonacci::Goal();
+
```

Comment:
This never actually 1) cancels a goal or 2) tests the behavior tree node abstraction you added above. That or any derivative of it needs to be inputted into the behavior tree to even call cancel and exercise the code you added. That's what we need test coverage over. We know that ROS action servers work by this point in the test coverage. Its about exercising and making sure the specific code you added is functioning properly

Reply:
okay, gottcha. Let me come up with a better one.

Reply:
Ah, I already have my cancel control node and a test for it. 

> so that your tests have tangible complete nodes to test in the BT itself.

Does that mean we can simply avoid avoid a test node for the abstract class? 

BTW, the abstract class test was just inspired from the `test_bt_action_node`. Maybe I just misunderstood the test coverage as you said.

Reply:
Yeah, if the control node derives from this abstraction, then it would exercise this code

Reply:
But wait the Fibonacci node already derives from the abstraction. Maybe, the naming should have been `CancelFibonacci `

Let me quickly explain the steps I do in the test: 

* I created a derivative of the `bt_cancel_action_node` in a dummy `Fibonacci` BT node (maybe the naming confused). 
* Create a server for the Fibonacci action
* Create a BT consisting of dummy Fibonacci node, and sent a dummy goal through `async_send_goal()`
* Finally, I just execute the tick for the BT consisting of the  Fibonacci node derived from  `bt_cancel_action_node` then we just cancel the goal through the Fibonacci node (again I named it wrongly, it should have been something like `CancelFibonacci`). Then for the test, I just check if the node returns `SUCCESS` 

We have the abstraction derived right? Sorry for troubling you again. Just wanted to clarify, so that I can make much better tests for someother issue, someother time.